README.md
# Gofile

[![GoDoc](https://godoc.org/github.com/siadat/gofile/http?status.svg)](https://godoc.org/github.com/siadat/gofile/http)

A non-blocking directory listing and file server.
It implementats HTTP/1.1 keepalive, chunked transfer, and byte range.

The HTTP server implementation provides a channel for writing chunked response. It could be used as a library. [Read the API](https://godoc.org/github.com/siadat/gofile/http).

![gofile](/../screenshots/screenshot-0.1.0.png?raw=true "gofile")

### Usage

    Usage: gofile [-v] <port> [<root>]

Examples:

    gofile 8080
    gofile 8080 ~/public

### Install

    go get -u github.com/siadat/gofile

### HTTP/1.1 implementation checklist

- [x] GET and HEAD methods
- [x] Support keep-alive connections
- [x] Support chunked transfer encoding
- [x] Requests must include a `Host` header
- [x] Requests with `Connection: close` should be closed
- [x] Support for requests with absolute URLs
- [x] If-Modified-Since support
- [x] Byte range support
- [ ] Transparent response compression

### Hacking

Submit an issue or send a pull request.
Make sure you `./run-tests.bash` to test your patch.

### Thanks

Thanks @valyala for his feature suggestions.


CHANGELOG.md
## 2016-07-21

- **fileserver** fileserver tries to open the following files before responding
  with 404:

    $uri
    $uri.html
    $uri.htm
    $uri/index.html
    $uri/index.htm


test.bash
host=127.0.0.1
port=8080
testdir="test-fixtures"

CRLF="\r\n"
HTTP11="HTTP/1.1${CRLF}"
HConn="Connection: close${CRLF}"
HHost="Host: localhost${CRLF}"
HIfModifiedPast="If-Modified-Since: $(gdate +"%a, %d %b %Y %T %Z" --date='@1000000000')${CRLF}"
HIfModifiedFuture="If-Modified-Since: $(gdate +"%a, %d %b %Y %T %Z" --date='@9000000000')${CRLF}"
HByteRange1="Range: bytes=-10${CRLF}"
HByteRange2="Range: bytes=90-${CRLF}"
HByteRange3="Range: bytes=10-10${CRLF}"
HByteRange4="Range: bytes=10-20${CRLF}"
HByteRange5="Range: bytes=-1${CRLF}"
HByteRange6="Range: bytes=0-10,20-${CRLF}" # Not supported

sendreq() {
	local ColorReq="" # \e[94m"
	local ColorNon="" # \e[0m"
	echo # "------------------------------------"
	echo -ne "${ColorReq}"
	echo -e "$1"
	echo -ne "${ColorNon}"
	echo -e "$1" | nc $host $port | head -n 15
}

hl() {
	echo
	echo "$1"
	echo "===================================="
}

hl "Valid: relative url => 200"
sendreq "GET / ${HTTP11}${HHost}${HConn}"
sendreq "HEAD / ${HTTP11}${HHost}${HConn}"

hl "Valid: modified, normal response => 200"
sendreq "GET /${testdir}/date.txt ${HTTP11}${HHost}${HIfModifiedPast}${HConn}"

hl "Valid: not modified, body should be empty => 304"
sendreq "GET /${testdir}/date.txt ${HTTP11}${HHost}${HIfModifiedFuture}${HConn}"

hl "Valid: range header => 200"
sendreq "GET /${testdir}/100.txt ${HTTP11}${HHost}${HByteRange1}${HConn}"
sendreq "GET /${testdir}/100.txt ${HTTP11}${HHost}${HByteRange2}${HConn}"
sendreq "GET /${testdir}/100.txt ${HTTP11}${HHost}${HByteRange3}${HConn}"
sendreq "GET /${testdir}/100.txt ${HTTP11}${HHost}${HByteRange4}${HConn}"
sendreq "GET /${testdir}/100.txt ${HTTP11}${HHost}${HByteRange5}${HConn}"
sendreq "GET /${testdir}/100.txt ${HTTP11}${HHost}${HByteRange6}${HConn}"

hl "Valid: absolute url => 200"
sendreq "GET http://localhost:8080/ ${HTTP11}${HHost}${HConn}"

hl "Valid: should not be chunked, no Transfer-Encoding header, must have Content-Length => 200"
sendreq "GET /${testdir}/date.txt ${HTTP11}${HHost}${HConn}"

hl "Valid: special characters => 200"
sendreq "GET /${testdir}/a+b+c+(d)/یک.txt ${HTTP11}${HHost}${HConn}"
sendreq "GET /${testdir}/a+b+c+(d)/e+f+g+[h]/test.txt ${HTTP11}${HHost}${HConn}"

hl "Valid: should not found => 404"
sendreq "GET /foo ${HTTP11}${HHost}${HConn}"

hl "Invalid: no 'Host' header => 400"
sendreq "GET / ${HTTP11}${HConn}"

hl "Invalid: bad paths => 401"
sendreq "GET ../ ${HTTP11}${HHost}${HConn}"
sendreq "GET /.. ${HTTP11}${HHost}${HConn}"
sendreq "GET http://localhost:8080/../ ${HTTP11}${HHost}${HConn}"

hl "Invalid: bad method => 501"
sendreq "POST / ${HTTP11}${HHost}${HConn}"

hl "A file larger than the response chunk size (1M) should be identical to the original file"
largefile="2m.file"
dd if=/dev/zero of=${testdir}/"$largefile" bs=2048 count=1024
wget -O "$largefile" "http://${host}:${port}/${testdir}/${largefile}"
diff -s "$largefile" ${testdir}/"$largefile"
rm "$largefile" ${testdir}/"$largefile"

hl "Valid: keepalive should not disconnect, because no '${HConn}' header is present"
sendreq "GET / ${HTTP11}${HHost}"


tests.bash
host=127.0.0.1
port=8080
testdir="test-fixtures"

CRLF="\r\n"
HTTP11="HTTP/1.1${CRLF}"
HTTPBAD="HTTP/BAD${CRLF}"
HConn="Connection: close${CRLF}"
HHost="Host: localhost${CRLF}"
HIfModifiedPast="If-Modified-Since: $(gdate +"%a, %d %b %Y %T %Z" --date='@1000000000')${CRLF}"
HIfModifiedFuture="If-Modified-Since: $(gdate +"%a, %d %b %Y %T %Z" --date='@9000000000')${CRLF}"
HByteRange1="Range: bytes=-10${CRLF}"
HByteRange2="Range: bytes=90-${CRLF}"
HByteRange3="Range: bytes=10-10${CRLF}"
HByteRange4="Range: bytes=10-20${CRLF}"
HByteRange5="Range: bytes=-1${CRLF}"
HByteRange6="Range: bytes=0-10,20-${CRLF}" # Not supported

sendreq() {
	local ColorReq="" # \e[94m"
	local ColorNon="" # \e[0m"
	echo # "------------------------------------"
	echo -ne "${ColorReq}"
	echo -e "$1"
	echo -ne "${ColorNon}"
	echo -e "$1" | nc $host $port # | head -n 15
}

hl() {
	echo
	echo "$1"
	echo "===================================="
}

hl "Valid: relative url => 200"
sendreq "GET / ${HTTP11}${HHost}${HConn}"
sendreq "HEAD / ${HTTP11}${HHost}${HConn}"

hl "Valid: modified, normal response => 200"
sendreq "GET /${testdir}/date.txt ${HTTP11}${HHost}${HIfModifiedPast}${HConn}"

hl "Valid: not modified, body should be empty => 304"
sendreq "GET /${testdir}/date.txt ${HTTP11}${HHost}${HIfModifiedFuture}${HConn}"

hl "Valid: range header => 200"
sendreq "GET /${testdir}/100.txt ${HTTP11}${HHost}${HByteRange1}${HConn}"
sendreq "GET /${testdir}/100.txt ${HTTP11}${HHost}${HByteRange2}${HConn}"
sendreq "GET /${testdir}/100.txt ${HTTP11}${HHost}${HByteRange3}${HConn}"
sendreq "GET /${testdir}/100.txt ${HTTP11}${HHost}${HByteRange4}${HConn}"
sendreq "GET /${testdir}/100.txt ${HTTP11}${HHost}${HByteRange5}${HConn}"
sendreq "GET /${testdir}/100.txt ${HTTP11}${HHost}${HByteRange6}${HConn}"

hl "Valid: absolute url => 200"
sendreq "GET http://localhost:8080/ ${HTTP11}${HHost}${HConn}"

hl "Valid: should not be chunked, no Transfer-Encoding header, must have Content-Length => 200"
sendreq "GET /${testdir}/date.txt ${HTTP11}${HHost}${HConn}"

hl "Valid: special characters => 200"
sendreq "GET /${testdir}/a+b+c+(d)/یک.txt ${HTTP11}${HHost}${HConn}"
sendreq "GET /${testdir}/a+b+c+(d)/e+f+g+[h]/test.txt ${HTTP11}${HHost}${HConn}"

hl "Valid: should not found => 404"
sendreq "GET /foo ${HTTP11}${HHost}${HConn}"

hl "Invalid: no 'Host' header => 400"
sendreq "GET / ${HTTP11}${HConn}"

hl "Invalid: bad paths => 401"
sendreq "GET ../ ${HTTP11}${HHost}${HConn}"
sendreq "GET /.. ${HTTP11}${HHost}${HConn}"
sendreq "GET http://localhost:8080/../ ${HTTP11}${HHost}${HConn}"

hl "Invalid: bad method => 501"
sendreq "POST / ${HTTP11}${HHost}${HConn}"

hl "Invalid: bad protocol"
sendreq "GET / ${HTTPBAD}${HHost}${HConn}"

hl "A file larger than the response chunk size (1M) should be identical to the original file"
largefile="2m.file"
dd if=/dev/zero of=${testdir}/"$largefile" bs=2048 count=1024
wget --quiet -O "$largefile" "http://${host}:${port}/${testdir}/${largefile}"
diff -s "$largefile" ${testdir}/"$largefile"
rm "$largefile" ${testdir}/"$largefile"

hl "Valid: keepalive should not disconnect, because no '${HConn}' header is present"
sendreq "GET / ${HTTP11}${HHost}" &
sleep 0.5
kill %1
wait


http/server_test.go
package http

import (
	"fmt"
	"testing"
)

// func TestServe() { }

func ExampleServe(t *testing.T) {
	Serve(8080, func(req Request, res *Response) {
		defer close(res.Body)
		res.Body <- []byte(fmt.Sprintf("You requested %v", req.URL))
	})
}


log/log.go
// Package log provides a simple interface for printing messages to stdout.
package log

import "fmt"

const (
	levelQuiet = iota
	levelNormal
	levelVerbose
)

var level = levelNormal

// SetQuiet sets the log level to quite. Only Error prints. All other logs are
// ignored.
func SetQuiet() {
	level = levelQuiet
}

// SetNormal sets the log level to normal. Info logs are ignored.
func SetNormal() {
	level = levelNormal
}

// SetVerbose sets the log level to verbose. All messages are printed.
func SetVerbose() {
	level = levelVerbose
}

// Info formats using the default formats for its operands and writes to
// standard output if level is levelVerbose.
func Info(a ...interface{}) {
	if level == levelVerbose {
		fmt.Println(a...)
	}
}

// Normal formats using the default formats for its operands and writes to
// standard output if level greater than or equal to levelNormal.
func Normal(a ...interface{}) {
	if level >= levelNormal {
		fmt.Println(a...)
	}
}

// Error formats using the default formats for its operands and writes to
// standard output, ignoring level.
func Error(a ...interface{}) {
	fmt.Println(a...)
}


fileserver.go
package main

import (
	"errors"
	"fmt"
	"io/ioutil"
	"mime"
	neturl "net/url"
	"os"
	"path"
	fp "path/filepath"
	"runtime"
	"strings"
	"time"

	http "github.com/siadat/gofile/http"
)

var (
	rootDir   = ""
	startTime = time.Now()
	history   []string
)

func htmlLayoutTop() string {
	return `<!DOCTYPE html>
    <html>
    <head>
      <title>Gofile</title>
      <meta charset="utf-8" />
	  <style>
	    body { font-family:monospace; padding:0; margin:0; }
	    h1,h2,h3,h4,h5,h6 { margin:10px; padding:10px; }
	    ul { padding:0; margin:0; }
	    li { padding:0; margin:1px 0; background-color:#eee; list-style:none; }
	    li.dir { background-color:#B8D2E8; font-weight:bold; /*eea;*/ }
	    li a { color:#29414E; text-decoration:none; display:block; padding:5px 20px; }
	  </style>
    </head>
    <body>`
}
func htmlLayoutBottom() string {
	return `</body></html>`
}

func htmlLink(href, text string) string {
	return fmt.Sprintf("<a href='%s'>%s</a>", href, text)
}

func filesLis(fileInfos []os.FileInfo, url string, urlUnescaped string, bodyChan chan []byte) {
	bodyChan <- []byte("<ul>")
	for _, fi := range fileInfos {
		filename := fi.Name()
		class := ""
		if fi.IsDir() {
			class = " class='dir' "
			filename = filename + "/"
		}

		fullPath := strings.Join([]string{url, neturl.QueryEscape(fi.Name())}, "/")
		// url could end with a "/" or with no "/", so when joined with
		// something else using "/" there could be a double slash ie "//"
		fullPath = strings.Replace(fullPath, "//", "/", 1)

		bodyChan <- []byte(fmt.Sprintf("<li%s>%s</li>\n", class, htmlLink(fullPath, filename)))
	}
	bodyChan <- []byte("</ul>")
	return
}

func listDirChan(url string, urlUnescaped string, filepath string, res *http.Response) (err error) {
	res.Body <- []byte(htmlLayoutTop())
	res.Body <- []byte(fmt.Sprintf("<h1>Directory Listing for %s</h1>", urlUnescaped))
	if false {
		res.Body <- []byte(fmt.Sprintf("<h3>Uptime:%s Goroutines:%d Requests:%d</h3>",
			time.Since(startTime),
			runtime.NumGoroutine(),
			len(history),
		))
	}

	fileInfos, status, errMsg := getFilesInDir(filepath)
	filesLis(fileInfos, url, urlUnescaped, res.Body)
	res.Body <- []byte(htmlLayoutBottom())

	if status != 200 {
		err = errors.New(errMsg)
		// content = ""
	}

	return
}

func getRootDir(optRoot string) (root string) {
	var wd string
	if len(optRoot) == 0 {
		wd, _ = os.Getwd()
	} else {
		wd = optRoot
	}

	root, err := fp.Abs(wd)

	if err != nil {
		panic(err)
	}
	return
}

func tryFilepaths(filepath string) (filepathRet string, file os.FileInfo, err error) {
	filepaths := []string{
		filepath,
		fmt.Sprintf("%s.html", filepath),
		fmt.Sprintf("%s.htm", filepath),
		fmt.Sprintf("%s/index.html", filepath),
		fmt.Sprintf("%s/index.htm", filepath),
	}

	for _, filepathRet = range filepaths {
		file, err = os.Stat(filepathRet)
		if err == nil {
			return filepathRet, file, nil
		}
	}

	return filepathRet, file, err
}

func urlToFilepath(requestURI string) (requestedFilepath string, retErr error) {
	// Note: path.Join removes '..', so the HasPrefix check is safe for paths
	// that try to traverse parent directory using '..'.
	if len(rootDir) == 0 {
		rootDir = getRootDir("")
	}
	requestedFilepath = path.Join(rootDir, requestURI)

	if !strings.HasPrefix(requestedFilepath, rootDir) {
		retErr = errors.New("Requested URI is not allowed")
		requestedFilepath = ""
		return
	}
	return
}

func getFilesInDir(requestedFilepath string) (fileInfos []os.FileInfo, status int, errMsg string) {
	status = 200
	fileInfos, err := ioutil.ReadDir(requestedFilepath)

	if err != nil {
		errMsg = "Requested URI was not found."
		status = 404
	}

	return
}

func getFilesize(filepath string) (fileSize int64) {
	f, err := os.Open(filepath)
	if err != nil {
		return
	}
	defer f.Close()

	if fi, err := f.Stat(); err == nil {
		fileSize = fi.Size()
	}
	return
}

func downloadFileChan(filepath string, ranges []http.ByteRange, res *http.Response) (err error) {
	// NOTE at the moment we are respecting the first range only
	rangeFrom := ranges[0].Start
	rangeTo := ranges[0].End

	f, err := os.Open(filepath)
	if err != nil {
		return
	}
	defer f.Close()

	var fileSize int64
	if fi, err := f.Stat(); err == nil {
		fileSize = fi.Size()
	}

	if rangeTo < 0 {
		rangeTo = fileSize + rangeTo
	}

	if rangeFrom < 0 {
		rangeFrom = fileSize + rangeFrom
	}

	rangeFromNew, err := f.Seek(rangeFrom, 0)
	if err != nil {
		f.Close()
		return
	}

	if rangeTo < rangeFromNew {
		rangeTo = rangeFromNew
	}

	var maxFileReadLen int64 = 1024 * 1024
	for cursorFrom := rangeFromNew; cursorFrom <= rangeTo; cursorFrom += maxFileReadLen {
		cursorTo := cursorFrom + maxFileReadLen - 1
		if cursorTo > rangeTo {
			cursorTo = rangeTo
		}
		buff := make([]byte, cursorTo-cursorFrom+1)

		_, err := f.Seek(cursorFrom, 0)
		if err != nil {
			break
		}

		readN, readErr := f.Read(buff)
		if readErr != nil {
			break
		}
		if readN == 0 {
			break
		}

		res.Body <- buff[:readN]
		// time.Sleep(time.Millisecond * 10)
	}

	return
}

func fileServerHandleRequestGen(optRoot string) func(http.Request, *http.Response) {
	rootDir = getRootDir(optRoot)
	return fileServerHandleRequest
}

func fileServerHandleRequest(req http.Request, res *http.Response) {
	history = append(history, req.URL.Path)

	unescapedURL, _ := neturl.QueryUnescape(req.URL.Path)
	filepath, err := urlToFilepath(unescapedURL)
	if err != nil {
		res.Status = 401
		defer close(res.Body)
		res.Body <- []byte(err.Error() + "\n")
		return
	}

	filepath, file, err := tryFilepaths(filepath)
	if err != nil {
		res.Status = 404
		defer close(res.Body)
		res.Body <- []byte("")
		return
	}

	if file.IsDir() {
		res.Status = 200
		res.ContentType = "text/html"
		defer close(res.Body)
		err = listDirChan(req.URL.Path, unescapedURL, filepath, res)
		// if err != nil { res.Status = 400 }
		return
	}

	res.Status = 200
	fileIsModified := true
	if len(req.Headers["If-Modified-Since"]) > 0 {
		ifModifiedSince := http.ParseHTTPDate(req.Headers["If-Modified-Since"])
		if !file.ModTime().After(ifModifiedSince) {
			fileIsModified = false
		}
	}

	if fileIsModified {
		res.ContentType = mime.TypeByExtension(fp.Ext(filepath))
		res.ContentLength = getFilesize(filepath)
		err = downloadFileChan(filepath, req.Ranges, res)
	} else {
		res.Status = 304
	}

	defer close(res.Body)

	if err != nil {
		res.Status = 400
	}
}


http/server.go
// Package http implements an HTTP/1.1 server.
package http

import (
	"fmt"
	"io"
	"math/rand"
	"net"
	"os"
	"strings"
	"time"

	log "github.com/siadat/gofile/log"
)

const (
	reqBuffLen    = 2 * 1024
	reqMaxBuffLen = 64 * 1024
)

var (
	socketCounter = 0
)

// Serve starts the HTTP server listening on port. For each request, handle is
// called with the parsed request and response in their own goroutine.
func Serve(port int, handle func(Request, *Response)) {
	tcpAddr, err := net.ResolveTCPAddr("tcp4", fmt.Sprintf(":%d", port))
	ln, err := net.ListenTCP("tcp", tcpAddr)
	if err != nil {
		log.Error(err)
		os.Exit(1)
		return
	}

	r := rand.New(rand.NewSource(99))

	for {
		conn, err := ln.Accept()
		if err != nil {
			log.Error("Error while accepting new connection", err)
			continue
		}

		socketCounter++
		log.Info("handleConnection #", socketCounter)
		req := Request{Headers: make(map[string]string)}
		res := Response{conn: conn, connID: r.Uint32()}
		go handleConnection(req, &res, handle)
	}
}

func readRequest(req Request, res *Response) (requestBuff []byte, err error) {
	requestBuff = make([]byte, 0, 8*1024)
	var reqLen int

	for {
		buff := make([]byte, reqBuffLen)
		reqLen, err = res.conn.Read(buff)
		requestBuff = append(requestBuff, buff[:reqLen]...)

		if len(requestBuff) > reqMaxBuffLen {
			log.Normal("Request is too big, ignoring the rest.")
			break
		}

		if err != nil && err != io.EOF {
			log.Error("Connection error:", err)
			break
		}

		if err == io.EOF || reqLen < reqBuffLen {
			break
		}
	}
	return
}

func handleConnection(req Request, res *Response, handle func(Request, *Response)) {
	defer func() {
		socketCounter--
		log.Info(fmt.Sprintf("Closing socket:%d. Total connections:%d", res.connID, socketCounter))
		// res.conn.Close()
		// FIXME this might be called before response is done writing
	}()

	for {
		requestBuff, err := readRequest(req, res)

		resStartTime := time.Now()

		if len(requestBuff) == 0 {
			return
		}

		if err != nil && err != io.EOF {
			log.Error("Error while reading socket:", err)
			return
		}

		log.Info(string(requestBuff[0:]))

		requestLines := strings.Split(string(requestBuff[0:]), crlf)
		req.parseHeaders(requestLines[1:])
		err = req.parseInitialLine(requestLines[0])

		res.Body = make(chan []byte)
		go res.respondOther(req)

		if err != nil {
			res.Status = 400
			res.ContentType = "text/plain"

			res.Body <- []byte(err.Error() + "\n")
			close(res.Body)

			// if req.Headers["Connection"] == "close" {
			// 	// res.conn.Close()
			// 	break
			// } else {
			// }
			continue
		}

		requestIsValid := true
		log.Normal(fmt.Sprintf("%s sock:%v %s",
			time.Now().Format("2006-01-02 15:04:05-0700"),
			res.connID,
			requestLines[0],
		))

		if len(req.Headers["Host"]) == 0 {
			res.ContentType = "text/plain"
			res.Status = 400
			close(res.Body)
			requestIsValid = false
			// continue
		}

		switch req.Method {
		case "GET", "HEAD":
		default:
			res.ContentType = "text/plain"
			res.Status = 501
			close(res.Body)
			requestIsValid = false
		}

		if requestIsValid {
			if req.Method == "HEAD" {
				close(res.Body)
			} else {
				handle(req, res)
			}
		}

		log.Normal(fmt.Sprintf("%s sock:%v Completed in %v",
			time.Now().Format("2006-01-02 15:04:05-0700"),
			res.connID,
			time.Since(resStartTime),
		))

		if req.Headers["Connection"] == "close" {
			break
		}
	}
}


http/request.go
package http

import (
	"errors"
	"fmt"
	"net/url"
	"strconv"
	"strings"
	"time"
)

// Protocol specifies the HTTP version, i.e., HTTP/1.1.
// Protocol string

type Request struct {
	// Method is the HTTP/1.1 method name.
	Method string

	// URL is the URL requested by the client.
	URL *url.URL

	// Headers contains pairs of the header key-values.
	Headers map[string]string

	// Ranges contains all the requested byte-ranges.
	Ranges []ByteRange

	isRanged bool
}

// ByteRange is the parsed ranges requested by the client.
// For example, the client could request the following byte range:
//
//     Range: bytes=100-
//
// This header is parsed with 100 and -1 (unspecified) being the start and end of the range.
type ByteRange struct {
	Start int64
	End   int64
}

// Length is the total number of bytes included in the range.
func (br ByteRange) Length() int64 {
	return br.End - br.Start + 1
}

func parseByteRangeHeader(headerValue string) (byteRanges []ByteRange, explicit bool) {
	rangePrefix := "bytes="
	byteRanges = make([]ByteRange, 0)

	if !strings.HasPrefix(headerValue, rangePrefix) {
		byteRanges = append(byteRanges, ByteRange{Start: 0, End: -1})
		return
	}

	explicit = true

	headerValue = headerValue[len(rangePrefix):]
	for _, value := range strings.Split(headerValue, ",") {

		// Let's say we have 10 bytes: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
		// -10 => [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
		// -7 => [3, 4, 5, 6, 7, 8, 9]
		// -2 => [8, 9]
		// -1 => [9]
		if val, err := strconv.ParseInt(value, 10, 0); err == nil && val < 0 {
			byteRanges = append(byteRanges, ByteRange{Start: val, End: -1})
			continue
		}

		// 0- => [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
		// 3- => [3, 4, 5, 6, 7, 8, 9]
		if strings.HasSuffix(value, "-") {
			if val, err := strconv.ParseInt(value[:len(value)-1], 10, 0); err == nil {
				byteRanges = append(byteRanges, ByteRange{Start: val, End: -1})
			}
			continue
		}

		// 1-1 => [1, 1]
		// 3-6 => [3, 4, 5, 6]
		rangeVals := strings.Split(value, "-")
		val1, err1 := strconv.ParseInt(rangeVals[0], 10, 0)
		val2, err2 := strconv.ParseInt(rangeVals[1], 10, 0)
		if err1 == nil && err2 == nil {
			byteRanges = append(byteRanges, ByteRange{Start: val1, End: val2})
		}
	}
	return
}

// ParseHTTPDate is a helper for parsing HTTP-dates.
func ParseHTTPDate(date string) (t time.Time) {
	t, err := time.Parse(httpTimeFormat, date)
	if err != nil {
		fmt.Println("error parsing", err)
	}
	return
}

func (req *Request) parseInitialLine(line string) (err error) {
	words := strings.SplitN(line, " ", 3)

	if len(words) < 3 {
		return errors.New("Invalid initial request line.")
	}

	if words[2] != "HTTP/1.1" {
		return errors.New("Invalid initial request line.")
	}

	req.Method = words[0]
	req.URL, _ = url.Parse(words[1])
	// req.Protocol = words[2]

	return
}

func (req *Request) parseHeaders(headerLines []string) {
	for _, headerLine := range headerLines {
		headerPair := strings.SplitN(headerLine, ": ", 2)
		if len(headerPair) == 2 {
			req.Headers[headerPair[0]] = headerPair[1]
		}
	}
	req.Ranges, req.isRanged = parseByteRangeHeader(req.Headers["Range"])
}


http/response.go
package http

import (
	"fmt"
	"net"
	"runtime"
	"strings"
	"time"

	log "github.com/siadat/gofile/log"
)

type Response struct {
	// Status is the response status code sent to the client.
	Status int

	// Body is the channel used by the server to write the content of the
	// response body. Each received array of byte is sent to the client as a HTTP
	// body chunk.
	Body chan []byte

	// ContentType is the meme type of the response. The default ContentType is
	// text/plain.
	ContentType string

	// ContentLength is the length of the response.
	ContentLength int64

	conn   net.Conn
	connID uint32
}

const (
	crlf           = "\r\n"
	httpTimeFormat = "Mon, 02 Jan 2006 15:04:05 MST"
	chunkLen       = 1024
	version        = "0.3.0"
)

var responsePhrases = map[int]string{
	100: "Continue",
	101: "Switching Protocols",
	200: "OK",
	201: "Created",
	202: "Accepted",
	203: "Non-Authoritative Information",
	204: "No Content",
	205: "Reset Content",
	206: "Partial Content",
	300: "Multiple Choices",
	301: "Moved Permanently",
	302: "Found",
	303: "See Other",
	304: "Not Modified",
	305: "Use Proxy",
	307: "Temporary Redirect",
	400: "Bad Request",
	401: "Unauthorized",
	402: "Payment Required",
	403: "Forbidden",
	404: "Not Found",
	405: "Method Not Allowed",
	406: "Not Acceptable",
	407: "Proxy Authentication Required",
	408: "Request Time-out",
	409: "Conflict",
	410: "Gone",
	411: "Length Required",
	412: "Precondition Failed",
	413: "Request Entity Too Large",
	414: "Request-URI Too Large",
	415: "Unsupported Media Type",
	416: "Requested range not satisfiable",
	417: "Expectation Failed",
	500: "Internal Server Error",
	501: "Not Implemented",
	502: "Bad Gateway",
	503: "Service Unavailable",
	504: "Gateway Time-out",
	505: "HTTP Version not supported",
}

func (res *Response) respondOther(req Request) {
	respond(req, res)
}

func respondHead(req Request, res *Response) {
	var headers []string

	if res.Status == 0 {
		res.Status = 200
	}

	if req.isRanged && res.Status == 200 {
		res.Status = 206
	}

	r := req.Ranges[0]
	if res.ContentLength > 0 {
		if r.End < 0 {
			r.End = res.ContentLength + r.End
		}
		if r.Start < 0 {
			r.Start = res.ContentLength + r.Start
		}
		if r.Start > r.End {
			res.Status = 416
		}
	}

	headers = append(headers, fmt.Sprintf("HTTP/1.1 %d %s", res.Status, responsePhrases[res.Status]))

	if req.isRanged && res.ContentLength > 0 {
		headers = append(headers, fmt.Sprintf("Content-Range: %s-%s/%d",
			fmt.Sprintf("%d", r.Start),
			fmt.Sprintf("%d", r.End),
			res.ContentLength))
	}

	if len(res.ContentType) == 0 {
		res.ContentType = "text/plain"
	}

	headers = append(headers,
		"Connection: keep-alive",
		"Accept-Ranges: byte",
		fmt.Sprintf("Content-Type: %s", res.ContentType),
		fmt.Sprintf("Server: Gofile/%s %s", version, runtime.Version()),
		fmt.Sprintf("Date: %s", time.Now().UTC().Format(httpTimeFormat)),
	)

	headers = append(headers, fmt.Sprintf("Transfer-Encoding: %s", "chunked"))

	if res.ContentLength > 0 {
		headers = append(headers, fmt.Sprintf("Content-Length: %d", r.Length()))
	}

	log.Info(strings.Join(headers, crlf) + crlf + crlf)
	res.conn.Write(([]byte)(strings.Join(headers, crlf) + crlf + crlf))
}

func respond(req Request, res *Response) {
	from := 0
	var chunkBuff []byte
	noWriteYet := true

	for content := range res.Body {
		if noWriteYet {
			noWriteYet = false
			respondHead(req, res)
			switch res.Status {
			case 304, 501:
				break
			}
		}

		if len(chunkBuff)+len(content) > chunkLen && len(chunkBuff) > 0 {
			to := from + len(chunkBuff)
			err := writeToConn(res.conn, chunkBuff, from, to)
			if err != nil {
				fmt.Println("Socket Write Error > ", err)
				break
			}
			from = 0
			chunkBuff = []byte{}
		}
		chunkBuff = append(chunkBuff, content...)
	}

	if len(chunkBuff) > 0 {
		writeToConn(res.conn, chunkBuff, 0, len(chunkBuff))
	}

	if noWriteYet {
		respondHead(req, res)
	} else {
		res.conn.Write(([]byte)(fmt.Sprintf("%d%s%s", 0, crlf, crlf)))
	}

	if req.Headers["Connection"] == "close" {
		res.conn.Close()
	}
}

func writeToConn(conn net.Conn, content []byte, from int, to int) (err error) {
	written := []byte(fmt.Sprintf("%x%s", to-from, crlf))
	written = append(written, content...)
	written = append(written, []byte(fmt.Sprintf("%s", crlf))...)
	_, err = conn.Write(written)
	return
}
